### PR TITLE
task : include userId in the log messages by using MDC.

### PR DESCRIPTION
**Pull request by mukeshk**
Addressed to #1102

### CODE REVIEW

**Comment by mukeshk** [comment]
To Discuss:
1. Package for the MdcLogEnhancerFilter 
2. URL Pattern for the Filter
3. Position of the UserId in the logged lines.
4. Few of the Requests are not able to show userId event after login and few are 
     a>  2019-07-27 20:09:28.148 <<UserID>>:<</UserID>> WARN 26456 --- [qtp985342789-28] r.m.web.feature.site.CspController
     b>2019-07-27 20:24:08.823 <<UserID>>:2<</UserID>> INFO 26456 --- [qtp985342789-28] r.m.w.f.c.CollectionServiceImpl          : Series #1 (stamps=1) has been added to collection of user #2

> While I am navigating after logging across pages, I am not able to get userId.


5. Temporary I have put a marker <userId>x</userId> in the logged.
    we can formulate how we want to represent.

**Comment by mukeshk** [comment]
Need your ideas.

**Comment by php-coder** [comment]
Awesome! I'll look in it later.

File: src/main/java/ru/mystamps/web/config/MdcLogEnhancerFilter.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	@Override
+	protected void doFilterInternal(HttpServletRequest httpServletRequest,
+									HttpServletResponse httpServletResponse,
+									FilterChain filterChain) throws ServletException, IOException {
+		String userId = "";
+		if (SecurityContextUtils.getUserId() != null) {
+			userId = SecurityContextUtils.getUserId().toString();
+		}
+		MDC.put("userId", "<<UserID>>:"+userId+"<</UserID>>");
+		filterChain.doFilter(httpServletRequest, httpServletResponse);
```
Comment (php-coder) [inline src/main/java/ru/mystamps/web/config/MdcLogEnhancerFilter.java]: We should also remove this attribute from MDC after finishing request processing. Use `MDC.remove()`.

Also, we should wrap doFilter() in catch block and do cleanup in "finally" block to ensure that we always clean after ourself.

File: src/main/java/ru/mystamps/web/config/MdcLogEnhancerFilter.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class MdcLogEnhancerFilter extends OncePerRequestFilter {
+	@Override
+	protected void doFilterInternal(HttpServletRequest httpServletRequest,
+									HttpServletResponse httpServletResponse,
+									FilterChain filterChain) throws ServletException, IOException {
+		String userId = "";
```
Comment (php-coder) [inline src/main/java/ru/mystamps/web/config/MdcLogEnhancerFilter.java]: This code should be simplified to somehting like this:
```java
Integer userId = ...;
if (userId != null) {
    MDC.put(...);
}
```

File: src/main/resources/application.properties [hunk 3]
```diff
@@ -45,3 +45,7 @@ app.mail.robot.email: dont-reply@my-stamps.ru
 # max time for reading the content will be also 1 sec. A timeout of zero is
 # interpreted as an infinite timeout.
 app.downloader.timeout: 2000
+
+# Setup log entry pattern to position the MDC userId parameter
+# Render the MDC data variable userId just before the priority field in the log output
+logging.pattern.level = %X{userId}%5p
```
Comment (php-coder) [inline src/main/resources/application.properties]: 1) let's add a link to a comment: https://docs.spring.io/spring-boot/docs/1.5.x/reference/html/boot-features-logging.html#boot-features-custom-log-configuration
2) we don't need the comment with explanation :)
3) let's try to use `[user:%X{userId}] %5p` pattern
4) please, move this property before app.mail

**Comment by mukeshk** [comment]
`2019-07-28 06:12:13.897 2  INFO 25796 --- [tp1370384110-30] m.w.f.i.DatabaseImagePersistenceStrategy : Image #1: preview not found`

`2019-07-28 06:13:19.085 2  INFO 25796 --- [tp1370384110-34] r.m.w.f.c.CollectionServiceImpl          : Series #1 (stamps=1) has been added to collection of user #2`

`2019-07-28 06:14:28.552 2  INFO 25796 --- [tp1370384110-35] r.m.w.f.c.CollectionServiceImpl          : Series #1 has been removed from collection of user #2`

Things which go through CSP - dont have it.
`2019-07-28 06:15:51.603   WARN 25796 --- [tp1370384110-33] r.m.web.feature.site.CspController`

How do I test it further? Check for GET/CREATE/REMOVE actions.

File: src/main/java/ru/mystamps/web/config/MvcConfig.java [hunk 4]
```diff
@@ -199,5 +200,13 @@ private static HandlerInterceptor getLocaleChangeInterceptor() {
 	private HandlerInterceptor getDownloadImageInterceptor() {
 		return new DownloadImageInterceptor(servicesConfig.getImageDownloaderService());
 	}
-	
+
+	@Bean
+	public FilterRegistrationBean getMdcLogEnhancerFilter() {
```
[the remark this answers is not in the document]
Comment (mukeshk) [reply src/main/java/ru/mystamps/web/config/MvcConfig.java]: changed to userMdcLogEnhancerFilter. 
Is this good ?

[the remark this answers is not in the document]
Comment (mukeshk) [reply src/main/java/ru/mystamps/web/config/MvcConfig.java]: Tried @Bean, also Tried @Filter - did not work.

[the remark this answers is not in the document]
Comment (php-coder) [reply src/main/java/ru/mystamps/web/config/MvcConfig.java]: Thanks! I'll take a look tonight.

[the remark this answers is not in the document]
Comment (php-coder) [reply src/main/java/ru/mystamps/web/config/MvcConfig.java]: Update: tomorrow :/

File: src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+									HttpServletResponse response,
+									FilterChain filterChain) throws ServletException, IOException {
+		Integer userId = SecurityContextUtils.getUserId();
+		if (userId != null) {
+			MDC.put("userId", userId.toString());
+		}
+		try {
+			filterChain.doFilter(request, response);
+		} finally {
+			MDC.remove("userId");
```
Comment (php-coder) [inline src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java]: Let's extract "userId" to a constant.

File: src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import java.io.IOException;
+
+public class UserMdcLoggingFilter extends OncePerRequestFilter {
+	@Override
+	protected void doFilterInternal(HttpServletRequest request,
+									HttpServletResponse response,
+									FilterChain filterChain) throws ServletException, IOException {
+		Integer userId = SecurityContextUtils.getUserId();
+		if (userId != null) {
+			MDC.put("userId", userId.toString());
```
Comment (php-coder) [inline src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java]: No need `toString()` here.

File: src/main/resources/application.properties [hunk 7]
```diff
@@ -36,6 +36,9 @@ server.compression.min-response-size: 512
 
 security.filter-dispatcher-types: REQUEST, ERROR
 
+# See for details: https://docs.spring.io/spring-boot/docs/1.5.x/reference/html/boot-features-logging.html#boot-features-custom-log-configuration
+logging.pattern.level = %X{userId} %5p
```
Comment (php-coder) [inline src/main/resources/application.properties]: AFAIR, I suggested to also add square brackets around: `[%X{userId}] %5p`

Comment (php-coder) [reply src/main/resources/application.properties]: Please, use `:` instead of `=` for separating key/values (like other properties do).

Comment (php-coder) [reply src/main/resources/application.properties]: I see that I suggested the following:

> let's try to use `[user:%X{userId}] %5p` pattern

File: src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class UserMdcLoggingFilter extends OncePerRequestFilter {
+	@Override
+	protected void doFilterInternal(HttpServletRequest request,
+									HttpServletResponse response,
+									FilterChain filterChain) throws ServletException, IOException {
```
Comment (php-coder) [inline src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java]: Why these 2 lines indented so heavily? Please, reduce (see how similar code is formatted).

**Comment by php-coder** [comment]
Could you, please, modify commit message?
```
task: include userId in the log messages by using MDC.

Related links:
- http://logback.qos.ch/manual/mdc.html
- http://docs.spring.io/spring-boot/docs/1.5.x/reference/html/boot-features-logging.html

Part of #388
Fix #1102
```

File: src/main/java/ru/mystamps/web/support/spring/security/SecurityConfig.java [hunk 9]
```diff
@@ -202,5 +202,14 @@ public FilterRegistrationBean getResetLocaleFilter(
 		
 		return bean;
 	}
-	
+
+	@Bean
+	public FilterRegistrationBean userMdcLogEnhancerFilter() {
```
Comment (php-coder) [inline src/main/java/ru/mystamps/web/support/spring/security/SecurityConfig.java]: userMdcLogEnhancerFilter  -> userMdcLoggingFilter

Comment (php-coder) [inline src/main/java/ru/mystamps/web/support/spring/security/SecurityConfig.java]: Locally it works for me if we just return instance of UserMdcLoggingFilter. Why it doesn't work to you? :/

File: src/main/resources/application.properties (hunk 7 above)
Comment (php-coder) [reply src/main/resources/application.properties]: Here is the final (?) pattern that I'd like to use: `[user:%-2X{userId}] %5p`

File: src/main/resources/application.properties [hunk 10]
```diff
@@ -36,6 +36,9 @@ server.compression.min-response-size: 512
 
 security.filter-dispatcher-types: REQUEST, ERROR
 
+# See for details: https://docs.spring.io/spring-boot/docs/1.5.x/reference/html/boot-features-logging.html#boot-features-custom-log-configuration
```
Comment (php-coder) [inline src/main/resources/application.properties]: Please, add yet another link here -- https://logback.qos.ch/manual/layouts.html

File: src/main/java/ru/mystamps/web/support/spring/security/SecurityConfig.java (hunk 9 above)
Comment (mukeshk) [reply src/main/java/ru/mystamps/web/support/spring/security/SecurityConfig.java]: I was trying to set it up without FilterRegistrationBean concept.
Directly annotating - @Filter or @Bean - Assuming spring will recognize as Filter and do the needful. 
Removed the @Bean annotation.

File: src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java (hunk 6 above)
Comment (mukeshk) [reply src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java]: The MDC.put force key: String, and Value: String

File: src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java (hunk 8 above)
Comment (mukeshk) [reply src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java]: formatted - looking at the other code.

File: src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java (hunk 6 above)
Comment (php-coder) [reply src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java]: Ah, got it. You're right. Let's use `String.valueOf(userId)` then.

Comment (php-coder) [reply src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java]: Ok, Let's leave it as-is then. Sorry and thank you! :)

File: src/main/java/ru/mystamps/web/support/spring/security/SecurityConfig.java [hunk 11]
```diff
@@ -203,4 +203,12 @@ public FilterRegistrationBean getResetLocaleFilter(
 		return bean;
 	}
 	
+	public FilterRegistrationBean userMdcLoggingFilter() {
```
Comment (php-coder) [inline src/main/java/ru/mystamps/web/support/spring/security/SecurityConfig.java]: I think that it won't work without `@Bean` annotation.

Comment (php-coder) [reply src/main/java/ru/mystamps/web/support/spring/security/SecurityConfig.java]: I'll try again tonight but I don't see why the following won't work:
```java
@Bean userMdcLoggingFilter() {
    return new UserMdcLoggingFilter();
}
```

File: src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+public class UserMdcLoggingFilter extends OncePerRequestFilter {
+	private static final String USER_ID="userId";
+	
+	@Override
+	protected void doFilterInternal(
+		HttpServletRequest request,
+		HttpServletResponse response,
+		FilterChain filterChain) 
+		throws ServletException, IOException {
+		Integer userId = SecurityContextUtils.getUserId();
```
Comment (php-coder) [inline src/main/java/ru/mystamps/web/support/spring/security/UserMdcLoggingFilter.java]: Could you, please, add a new line before  this line? To separate a method body from its parameters.

File: src/main/resources/application.properties (hunk 7 above)
Comment (php-coder) [reply src/main/resources/application.properties]: ^^^ this is a pattern that we'll use. Could you update?

**Comment by php-coder** [comment]
The first line of a commit message still has to be updated to

> task: include userId in the log messages by using MDC.

**Comment by mukeshk** [comment]
please review.
I have set the pattern like :
logging.pattern.level: [user:%-2X{userId}] %5p

Is this good?

**Comment by php-coder** [comment]
> Is this good?

Yes, thank you! Could you update git commit also? (see my comments above).

**Comment by php-coder** [comment]
There is an issue with a current implementation: this filter doesn't get applied to error pages (try to open `//test` and you see that userId is always empty). I haven't try it yet, but it seems like we should use [`setDispatcherType()` method](https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/web/servlet/AbstractFilterRegistrationBean.html#setDispatcherTypes-javax.servlet.DispatcherType-javax.servlet.DispatcherType...-). We need to bind to REQUEST and ERROR dispatch types.

Yes, it means that we need to put back FilterRegistrationBean back :)

**Comment by mukeshk** [comment]
I have put the @Bean public FilterRegistrationBean userMdcLoggingFilter() {
still for few request user is not recorded.

2019-07-29 20:34:53.224 [user:  ]  WARN 23944 --- [tp1635327890-33] r.m.web.feature.site.CspController
2019-07-29 20:35:18.972 [user:  ]  WARN 23944 --- [tp1635327890-31] o.eclipse.jetty.servlet.ServletHandler   : //test

Other requests like adding, and remove collection - event - userId is recorded.

**Comment by php-coder** [comment]
> Things which go through CSP - dont have it.

I've played with it and here is what I found out: it works when `report-uri` in CSP header matches with a site that is open in a browser. When you run the app locally, report-uri is set to `http://127.0.0.1:8080`, so if you open this page in a browser you will see that userId is logged. But it won't work if you open `http://locahost:8080`.

**Comment by php-coder** [comment]
> There is an issue with a current implementation: this filter doesn't get applied to error pages (try to open //test and you see that userId is always empty). I haven't try it yet, but it seems like we should use setDispatcherType() method. We need to bind to REQUEST and ERROR dispatch types.

I've created an issue for tracking it:  #1106

**Comment by php-coder** [comment]
@mukeshk Good work, thank you! 👍

**Pull request merged**